src/smt/quant_elim_solver.cpp

```diff
@@ -0,0 +1,106 @@
+/*********************                                                        */
+/*! \file quant_elim_solver.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The solver for quantifier elimination queries
+ **/
+
+#include "smt/quant_elim_solver.h"
+
+#include "smt/smt_solver.h"
+#include "theory/quantifiers/extended_rewrite.h"
+#include "theory/rewriter.h"
+#include "theory/theory_engine.h"
+
+using namespace CVC4::theory;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace smt {
+
+QuantElimSolver::QuantElimSolver(SmtSolver& sms) : d_smtSolver(sms) {}
+
+QuantElimSolver::~QuantElimSolver() {}
+
+Node QuantElimSolver::getQuantifierElimination(Assertions& as,
+                                               Node q,
+                                               bool doFull)
+{
+  Trace("smt-qe") << "Do quantifier elimination " << q << std::endl;
+  if (q.getKind() != EXISTS && q.getKind() != FORALL)
+  {
+    throw ModalException(
+        "Expecting a quantified formula as argument to get-qe.");
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  // tag the quantified formula with the quant-elim attribute
+  TypeNode t = nm->booleanType();
+  Node n_attr = nm->mkSkolem("qe", t, "Auxiliary variable for qe attr.");
+  std::vector<Node> node_values;
+  TheoryEngine* te = d_smtSolver.getTheoryEngine();
+  Assert(te != nullptr);
+  te->setUserAttribute(
+      doFull ? "quant-elim" : "quant-elim-partial", n_attr, node_values, "");
+  n_attr = nm->mkNode(INST_ATTRIBUTE, n_attr);
+  n_attr = nm->mkNode(INST_PATTERN_LIST, n_attr);
+  std::vector<Node> children;
+  children.push_back(q[0]);
+  children.push_back(q.getKind() == EXISTS ? q[1] : q[1].negate());
+  children.push_back(n_attr);
+  Node ne = nm->mkNode(EXISTS, children);
+  Trace("smt-qe-debug") << "Query for quantifier elimination : " << ne
+                        << std::endl;
+  Assert(ne.getNumChildren() == 3);
+  // We consider this to be an entailment check, which also avoids incorrect
+  // status reporting (see SmtEngineState::d_expectedStatus).
+  Result r =
+      d_smtSolver.checkSatisfiability(as, std::vector<Node>{ne}, false, true);
```

Comment:
Here, you changed the `inUnsatCore` flag to `false`. Just to make sure that you intended to do that...

Reply:
Yes, it made no sense why inUnsatCore would be true.  I've also commented on why the other flag is true (which was also strange to me).